### PR TITLE
Added new object vzRsSubjFiltAtt [aci_contract_subject_filter]

### PR DESCRIPTION
**Pull request by anvitha-jain**

### CODE REVIEW

File: aci/data_source_aci_vzsubj.go [hunk 1]
```diff
@@ -62,6 +63,44 @@ func dataSourceAciContractSubject() *schema.Resource {
...
+			},
+			"consumer_to_provider": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				// DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+				// 	if (k == "consumer_to_provider.target_dscp" || k == "consumer_to_provider.prio") && new != old {
+				// 		return false
+				// 	}
+				// 	return true
+				// },
```
Comment (lhercot) [inline aci/data_source_aci_vzsubj.go:84]: remove if not needed

File: aci/data_source_aci_vzsubj.go [hunk 2]
```diff
@@ -62,6 +63,44 @@ func dataSourceAciContractSubject() *schema.Resource {
...
+			},
+			"provider_to_consumer": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				// DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+				// 	if (k == "provider_to_consumer.target_dscp" || k == "provider_to_consumer.prio") && new != old {
+				// 		return false
+				// 	}
+				// 	return true
+				// },
```
Comment (lhercot) [inline aci/data_source_aci_vzsubj.go:98]: remove if not needed

File: aci/data_source_aci_vzsubj.go [hunk 3]
```diff
@@ -62,6 +63,44 @@ func dataSourceAciContractSubject() *schema.Resource {
 				Optional: true,
 				Computed: true,
 			},
+			"apply_both_directions": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
```
Comment (lhercot) [inline aci/data_source_aci_vzsubj.go:69]: remove

File: aci/data_source_aci_vzsubj.go [hunk 4]
```diff
@@ -62,6 +63,44 @@ func dataSourceAciContractSubject() *schema.Resource {
...
+				Default:  "yes",
```
Comment (lhercot) [inline aci/data_source_aci_vzsubj.go:70]: remove

File: aci/resource_aci_vzrssubjfiltatt.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+				Optional: true,
+				Computed: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"default",
+					"level1",
+					"level2",
+					"level3",
+				}, false),
+			},
+			"tn_vz_filter_name": {
```
Comment (lhercot) [inline aci/resource_aci_vzrssubjfiltatt.go:69]: filter_dn

File: aci/resource_aci_vzsubj.go [hunk 6]
```diff
@@ -140,6 +151,34 @@ func resourceAciContractSubject() *schema.Resource {
...
+			"consumer_to_provider": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					if (k == "consumer_to_provider.target_dscp" || k == "consumer_to_provider.prio") && new != old {
+						return false
+					}
+					return true
+				},
+				Description: "Create InTerm",
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:163]: Set InTerm attributes

File: aci/resource_aci_vzsubj.go [hunk 7]
```diff
@@ -140,6 +151,34 @@ func resourceAciContractSubject() *schema.Resource {
...
+			"provider_to_consumer": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					if (k == "provider_to_consumer.target_dscp" || k == "provider_to_consumer.prio") && new != old {
+						return false
+					}
+					return true
+				},
+				Description: "Create OutTerm",
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:177]: Set OutTerm attributes

File: aci/resource_aci_vzsubj.go [hunk 8]
```diff
@@ -353,6 +483,42 @@ func resourceAciContractSubjectUpdate(ctx context.Context, d *schema.ResourceDat
 		return diag.FromErr(err)
 	}
 
+	if d.HasChange("apply_both_directions") || d.HasChange("consumer_to_provider") || d.HasChange("provider_to_consumer") {
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:486]: Don't think you need d.HasChange("apply_both_directions") anymore.

Comment (lhercot) [reply aci/resource_aci_vzsubj.go:486]: if apply_both_directions == "no" && (d.HasChange("consumer_to_provider") || d.HasChange("provider_to_consumer"))

File: examples/aci_contract_subject_filter/main.tf [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  tenant_dn                = aci_tenant.tenant_for_contract_filter.id
+  name                     = "test_tf_contract"
+}
+
+resource "aci_contract_subject" "contract_subject" {
+  contract_dn   = aci_contract.test_contract.id
+  name          = "contract_subject"
+  rev_flt_ports = "no"
+}
+
```
Comment (lhercot) [inline examples/aci_contract_subject_filter/main.tf:31]: add aci_filter resource

File: examples/aci_contract_subject_filter/main.tf [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+resource "aci_contract_subject_filter" "contract_subject_filter" {
+  contract_subject_dn = aci_contract_subject.contract_subject.id
+  action = "permit"
+  directives = ["log"]
+  priority_override = "default"
+  tn_vz_filter_name = "test_filter"
```
Comment (lhercot) [inline examples/aci_contract_subject_filter/main.tf:37]: filter_dn = aci_filter.myfilter.id

File: examples/aci_contract_subject_filter/main.tf [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+data "aci_contract_subject_filter" "example" {
+  contract_subject_dn  = aci_contract_subject_filter.contract_subject_filter.contract_subject_dn
+  tn_vz_filter_name  = aci_contract_subject_filter.contract_subject_filter.tn_vz_filter_name
```
Comment (lhercot) [inline examples/aci_contract_subject_filter/main.tf:42]: filter_dn = aci_filter.myfilter.id

File: examples/contract_subject/main.tf [hunk 12]
```diff
@@ -12,4 +12,3 @@ provider "aci" {
   url      = ""
   insecure = true
 }
-
```
Comment (lhercot) [inline examples/contract_subject/main.tf:15]: add back empty line

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/data_source_aci_vzrssubjfiltatt.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+			},
+			"priority_override": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			// "tn_vz_filter_name": {
+			// 	Type:     schema.TypeString,
+			// 	Required: true,
+			// },
```
Comment (lhercot) [inline aci/data_source_aci_vzrssubjfiltatt.go:43]: remove this

File: aci/resource_aci_vzrssubjfiltatt.go [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+		}
+	} else {
+		d.Set("directives", directivesGet)
+	}
+	d.Set("priority_override", vzRsSubjFiltAttMap["priorityOverride"])
+	tnVzFilterName := GetMOName(d.Get("filter_dn").(string))
+	if tnVzFilterName != vzRsSubjFiltAttMap["tnVzFilterName"] {
+		d.Set("filter_dn", "")
+	}
+	d.Set("tn_vz_filter_name", vzRsSubjFiltAttMap["tnVzFilterName"])
```
Comment (lhercot) [inline aci/resource_aci_vzrssubjfiltatt.go:125]: why set tn_vz_filter_name ? We should set filter_dn to tDn

File: aci/resource_aci_vzrssubjfiltatt.go [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+func resourceAciSubjectFilterImport(d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
+	log.Printf("[DEBUG] %s: Beginning Import", d.Id())
+	aciClient := m.(*client.Client)
+	dn := d.Id()
+	vzRsSubjFiltAtt, err := getRemoteSubjectFilter(aciClient, dn)
+	if err != nil {
+		return nil, err
+	}
+	schemaFilled, err := setSubjectFilterAttributes(vzRsSubjFiltAtt, d)
```
Comment (lhercot) [inline aci/resource_aci_vzrssubjfiltatt.go:137]: setSubjectFilterAttributes uses d.GetOk or d.Get which are not available in import how is it still properly imported?

File: aci/resource_aci_vzsubj.go [hunk 16]
```diff
@@ -179,10 +218,72 @@ func setContractSubjectAttributes(vzSubj *models.ContractSubject, d *schema.Reso
 	d.Set("prio", vzSubjMap["prio"])
 	d.Set("prov_match_t", vzSubjMap["provMatchT"])
 	d.Set("rev_flt_ports", vzSubjMap["revFltPorts"])
+	d.Set("apply_both_directions", vzSubjMap["applyBothDirections"])
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:221]: do you actually get it from the API?

File: aci/resource_aci_vzsubj.go [hunk 17]
```diff
@@ -243,13 +344,88 @@ func resourceAciContractSubjectCreate(ctx context.Context, d *schema.ResourceDat
 	if TargetDscp, ok := d.GetOk("target_dscp"); ok {
 		vzSubjAttr.TargetDscp = TargetDscp.(string)
 	}
+	ApplyBothDirections := d.Get("apply_both_directions")
+
+	if ApplyBothDirections == "yes " {
+		if (d.Get("consumer_to_provider") != nil || len(d.Get("consumer_to_provider").(map[string]interface{})) != 0) || (d.Get("provider_to_consumer") != nil || len(d.Get("provider_to_consumer").(map[string]interface{})) == 0) {
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:350]: both should be != 0

File: website/docs/d/contract_subject.html.markdown [hunk 18]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
 
 ## Argument Reference
 
-- `contract_dn` - (Required) Distinguished name of parent Contract object.
+- `contract_dn` - (Required) Distinguished name of parent (Contract) object.
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:25]: Why adding parenthesis?

File: website/docs/d/contract_subject.html.markdown [hunk 19]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
...
 - `name` - (Required) name of Object contract_subject.
 
 ## Attribute Reference
 
 - `id` - Attribute id set to the Dn of the Contract Subject.
-- `annotation` - (Optional) Annotation for object contract subject.
-- `description` - (Optional) Description for object contract subject.
+- `annotation` - (Optional) Annotation for the object contract subject.
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:31]: the contract subject object (do same for others in this documentation files)

File: website/docs/d/contract_subject.html.markdown [hunk 20]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
...
+- `description` - (Optional) Description for the object contract subject.
 - `cons_match_t` - (Optional) The subject match criteria across consumers.
-- `name_alias` - (Optional) Name alias for object contract subject.
+- `name_alias` - (Optional) Name alias for the object contract subject.
 - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
 - `prov_match_t` - (Optional) The subject match criteria across consumers.
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:36]: across providers

File: website/docs/d/contract_subject.html.markdown [hunk 21]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
...
 - `rev_flt_ports` - (Optional) Enables filter to apply on ingress and egress traffic.
 - `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
+
+- `apply_both_directions` - (Optional) . By default set to "yes".
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:40]: What does it do? Add that before the .

File: website/docs/d/contract_subject.html.markdown [hunk 22]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
...
+- ` consumer_to_provider` - (Optional) Filter Chain For Consumer to Provider. Class vzInTerm.
+    - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:42]: The priority level to assign to traffic matching the consumer to provider flows.

File: website/docs/d/contract_subject.html.markdown [hunk 23]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
...
+    - `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
+    - `relation_vz_rs_in_term_graph_att` - (Optional) Relation to class L4-L7 Service Graph (vnsAbsGraph).
+- `provider_to_consumer` - (Optional) Filter Chain For Provider to Consumer. Class vzOutTerm
+    - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:46]: The priority level to assign to traffic matching the provider to consumer flows.

File: website/docs/d/contract_subject.html.markdown [hunk 24]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
...
+- `name_alias` - (Optional) Name alias for the object contract subject.
 - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
 - `prov_match_t` - (Optional) The subject match criteria across consumers.
 - `rev_flt_ports` - (Optional) Enables filter to apply on ingress and egress traffic.
 - `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
+
+- `apply_both_directions` - (Optional) . By default set to "yes".
+- ` consumer_to_provider` - (Optional) Filter Chain For Consumer to Provider. Class vzInTerm.
+    - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
+    - `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:43]: The target DSCP to assign to traffic matching the consumer to provider flows.

File: website/docs/d/contract_subject.html.markdown [hunk 25]
```diff
@@ -22,17 +22,27 @@ data "aci_contract_subject" "dev_subject" {
...
+    - `relation_vz_rs_in_term_graph_att` - (Optional) Relation to class L4-L7 Service Graph (vnsAbsGraph).
+- `provider_to_consumer` - (Optional) Filter Chain For Provider to Consumer. Class vzOutTerm
+    - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
+    - `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
```
Comment (lhercot) [inline website/docs/d/contract_subject.html.markdown:47]: The target DSCP to assign to traffic matching the provider to consumer flows.

File: website/docs/r/contract_subject.html.markdown [hunk 26]
```diff
@@ -45,6 +69,17 @@ Manages ACI Contract Subject
 - `relation_vz_rs_sdwan_pol` - (Optional) Relation to class extdevSDWanSlaPol. Cardinality - N_TO_ONE. Type - String.
 - `relation_vz_rs_subj_filt_att` - (Optional) Relation to class vzFilter. Cardinality - N_TO_M. Type - Set of String.
 
+- `apply_both_directions` - (Optional) . By default set to "yes".
```
Comment (lhercot) [inline website/docs/r/contract_subject.html.markdown:72]: What is it used for?

File: website/docs/r/contract_subject.html.markdown [hunk 27]
```diff
@@ -45,6 +69,17 @@ Manages ACI Contract Subject
...
+- ` consumer_to_provider` - (Optional) Filter Chain For Consumer to Provider. Class vzInTerm.
+    - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
+    - `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
+	- `relation_vz_rs_in_term_graph_att` - (Optional) Relation to class L4-L7 Service Graph (vnsAbsGraph).
+- `provider_to_consumer` - (Optional) Filter Chain For Provider to Consumer. Class vzOutTerm
+    - `prio` - (Optional) The priority level of a sub application running behind an endpoint group, such as an Exchange server.
+    - `target_dscp` - (Optional) The target differentiated services code point (DSCP) of the path attached to the layer 3 outside profile.
+	- `relation_vz_rs_out_term_graph_att` - (Optional) Relation to class L4-L7 Service Graph (vnsAbsGraph).
+
```
Comment (lhercot) [inline website/docs/r/contract_subject.html.markdown:81]: see comments in data source.

File: website/docs/r/contract_subject_filter.html.markdown [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+## Example Usage ##
+
+```hcl
+resource "aci_contract_subject_filter" "example" {
+  contract_subject_dn  = aci_contract_subject.example.id
+  filter_dn  = aci_filter.example.id
+  action = "permit"
+  directives = ["none"]
+  priority_override = "default"
+
```
Comment (lhercot) [inline website/docs/r/contract_subject_filter.html.markdown:33]: remove empty line

File: website/docs/r/contract_subject_filter.html.markdown [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+}
+```
+
+## Argument Reference ##
+
+* `contract_subject_dn` - (Required) Distinguished name of the parent Contract Subject object.
+* `filter_dn` - (Required) Distinguished name of the Filter object.
+* `annotation` - (Optional) Annotation of the Contract Subject Filter object.
+* `action` - (Optional) The action required when the condition is met. Allowed values are "deny", "permit", and default value is "permit". Type: String.
```
Comment (lhercot) [inline website/docs/r/contract_subject_filter.html.markdown:42]: and the default value

File: website/docs/r/contract_subject_filter.html.markdown [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+* `directives` - (Optional) Directives of the Contract Subject Filter object. Allowed values are "log", "no_stats", "none", and default value is "none". Type: List.
```
Comment (lhercot) [inline website/docs/r/contract_subject_filter.html.markdown:43]: and the default value

File: website/docs/r/contract_subject_filter.html.markdown [hunk 31]
```diff
@@ -0,0 +1,55 @@
...
+* `priority_override` - (Optional) Priority override of the Contract Subject Filter object. Allowed values are "default", "level1", "level2", "level3", and default value is "default". Type: String.
```
Comment (lhercot) [inline website/docs/r/contract_subject_filter.html.markdown:44]: and the default value

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/resource_aci_vzsubj.go (hunk 16 above)
Comment (anvitha-jain) [reply aci/resource_aci_vzsubj.go:221]: no we don't get apply_both_directions from the API but if I set the parameter in setContractSubjectAttributes then I don't have to do it everytime i need to set parameters.

File: aci/resource_aci_vzsubj.go [hunk 32]
```diff
@@ -456,6 +1262,127 @@ func resourceAciContractSubjectRead(ctx context.Context, d *schema.ResourceData,
...
+			vzOutTermMap["relation_vz_rs_out_term_graph_att"] = vzRsOutTermGraphAttData.(string)
+		}
+
+		vzRsOutTermFilterAttData, err := aciClient.ListFilterRelationship(vzOutTerm.DistinguishedName)
+		if err != nil {
+			log.Printf("[DEBUG] Error while reading relation vzRsOutTermFilterAtt %v", err)
+		} else {
+			vzOutTermFilterSet := make([]interface{}, 0, 1)
+			for _, vzRsOutTermFilter := range vzRsOutTermFilterAttData {
+
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:1349]: Remove empty line

File: website/docs/r/contract_subject.html.markdown [hunk 33]
```diff
@@ -26,25 +26,72 @@ Manages ACI Contract Subject
...
 
 - `relation_vz_rs_subj_graph_att` - (Optional) Relation to class vnsAbsGraph. Cardinality - N_TO_ONE. Type - String.
 - `relation_vz_rs_sdwan_pol` - (Optional) Relation to class extdevSDWanSlaPol. Cardinality - N_TO_ONE. Type - String.
 - `relation_vz_rs_subj_filt_att` - (Optional) Relation to class vzFilter. Cardinality - N_TO_M. Type - Set of String.
 
+- `apply_both_directions` - (Optional) Defines if a contract subject allows traffic matching the defined filters for both consumer-to-provider and provider-to-consumer together or if each direction can be defined separately. By default set to "yes".
+  - When set to "yes", the filters defined in the subject are applied for both directions.
+  - When set to "no", the filters for each direction (consumer-to-provider and provider-to-consumer) are defined independently.
+- ` consumer_to_provider` - (Optional) Filter Chain for Consumer to Provider. Class vzInTerm.
```
Comment (lhercot) [inline website/docs/r/contract_subject.html.markdown:75]: Type Block

File: website/docs/r/contract_subject.html.markdown [hunk 34]
```diff
@@ -26,25 +26,72 @@ Manages ACI Contract Subject
...
+    - `prio` - (Optional) The priority level to assign to traffic matching the consumer to provider flows.
+    - `target_dscp` - (Optional) The target DSCP to assign to traffic matching the consumer to provider flows.
+	- `relation_vz_rs_in_term_graph_att` - (Optional) Relation to class L4-L7 Service Graph (vnsAbsGraph).
+	-`relation_vz_rs_filt_att` - (Optional) Relation to class Filters (vzRsFiltAtt).
+      - `action` - (Optional) The action required when the condition is met. Allowed values are "deny", "permit", and the default value is "permit".
+      - `directives` - (Optional) Directives of the Contract Subject Filter object for Consumer to Provider. Allowed values are "log", "no_stats", "none", and the default value is "none".
+      - `priority_override` - (Optional) Priority override of the Consumer to Provider Filter object. Allowed values are "default", "level1", "level2", "level3", and the default value is "default".
+      - `filter_dn` - (Required) Distinguished name of the Filter object for Consumer to Provider.
+- `provider_to_consumer` - (Optional) Filter Chain For Provider to Consumer. Class vzOutTerm
```
Comment (lhercot) [inline website/docs/r/contract_subject.html.markdown:84]: Type Block

File: website/docs/r/contract_subject.html.markdown [hunk 35]
```diff
@@ -26,25 +26,72 @@ Manages ACI Contract Subject
...
+    - `prio` - (Optional) The priority level to assign to traffic matching the provider to consumer flows.
+    - `target_dscp` - (Optional) The target DSCP to assign to traffic matching the provider to consumer flows.
+	- `relation_vz_rs_out_term_graph_att` - (Optional) Relation to class L4-L7 Service Graph (vnsAbsGraph).
+	-`relation_vz_rs_filt_att` - (Optional) (Optional) Relation to class Filters (vzRsFiltAtt).
+      - `action` - (Optional) The action required when the condition is met. Allowed values are "deny", "permit", and the default value is "permit".
```
Comment (lhercot) [inline website/docs/r/contract_subject.html.markdown:89]: Add that it is only used when action is deny

File: website/docs/r/contract_subject_filter.html.markdown [hunk 36]
```diff
@@ -0,0 +1,54 @@
...
+
+## Example Usage ##
+
+```hcl
+resource "aci_contract_subject_filter" "example" {
+  contract_subject_dn  = aci_contract_subject.example.id
+  filter_dn  = aci_filter.example.id
+  action = "permit"
+  directives = ["none"]
+  priority_override = "default"
```
Comment (lhercot) [inline website/docs/r/contract_subject_filter.html.markdown:32]: don't show an example that is incorrect. either use deny or remove this line.

File: website/docs/r/contract_subject_filter.html.markdown [hunk 37]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+## Argument Reference ##
+
+* `contract_subject_dn` - (Required) Distinguished name of the parent Contract Subject object.
+* `filter_dn` - (Required) Distinguished name of the Filter object.
+* `annotation` - (Optional) Annotation of the Contract Subject Filter object.
+* `action` - (Optional) The action required when the condition is met. Allowed values are "deny", "permit", and the default value is "permit". Type: String.
+* `directives` - (Optional) Directives of the Contract Subject Filter object. Allowed values are "log", "no_stats", "none", and the default value is "none". Type: List.
+* `priority_override` - (Optional) Priority override of the Contract Subject Filter object. Allowed values are "default", "level1", "level2", "level3", and the default value is "default". Type: String.
```
Comment (lhercot) [inline website/docs/r/contract_subject_filter.html.markdown:43]: Add that it is only used with action is deny

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/resource_aci_vzrssubjfiltatt.go [hunk 38]
```diff
@@ -0,0 +1,257 @@
...
+		vzRsSubjFiltAttAttr.Directives = Directives
+	}
+
+	if PriorityOverride, ok := d.GetOk("priority_override"); ok {
+		vzRsSubjFiltAttAttr.PriorityOverride = PriorityOverride.(string)
+	}
+
+	if TnVzFilterName, ok := d.GetOk("tn_vz_filter_name"); ok {
+		vzRsSubjFiltAttAttr.TnVzFilterName = TnVzFilterName.(string)
+	}
```
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply aci/resource_aci_vzrssubjfiltatt.go:163]: API needs filtername as part of payload this is inside create function

File: aci/resource_aci_vzrsfiltatt.go [hunk 39]
```diff
@@ -0,0 +1,261 @@
...
+		vzRsFiltAttAttr.Directives = Directives
+	}
+
+	if PriorityOverride, ok := d.GetOk("priority_override"); ok {
+		vzRsFiltAttAttr.PriorityOverride = PriorityOverride.(string)
+	}
+
+	if TnVzFilterName, ok := d.GetOk("tn_vz_filter_name"); ok {
+		vzRsFiltAttAttr.TnVzFilterName = TnVzFilterName.(string)
+	}
```
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply aci/resource_aci_vzrsfiltatt.go:212]: API needs filtername as part of payload this is inside create and update function

File: aci/resource_aci_vzrsfiltatt.go [hunk 40]
```diff
@@ -0,0 +1,261 @@
...
```
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply aci/resource_aci_vzrsfiltatt.go:167]: API needs filtername as part of payload this is inside create/update function

File: aci/resource_aci_vzrssubjfiltatt.go [hunk 41]
```diff
@@ -0,0 +1,257 @@
...
```
[the remark this answers is not in the document]
Comment (anvitha-jain) [reply aci/resource_aci_vzrssubjfiltatt.go:208]: API needs filtername as part of payload this is inside create/update function

File: aci/resource_aci_vzrsfiltatt.go (hunk 40 above)
[the remark this answers is not in the document]
Comment (lhercot) [reply aci/resource_aci_vzrsfiltatt.go:167]: but on line 137 you already set tnVzFilterName so here it should be 
```
vzRsFiltAttAttr.TnVzFilterName = tnVzFilterName
```

And no need for the if.

File: aci/resource_aci_vzrsfiltatt.go (hunk 39 above)
[the remark this answers is not in the document]
Comment (lhercot) [reply aci/resource_aci_vzrsfiltatt.go:212]: but on line 183 you already set tnVzFilterName so here it should be 
```
vzRsFiltAttAttr.TnVzFilterName = tnVzFilterName
```

And no need for the if.

File: aci/resource_aci_vzrssubjfiltatt.go (hunk 38 above)
[the remark this answers is not in the document]
Comment (lhercot) [reply aci/resource_aci_vzrssubjfiltatt.go:163]: but on line 133 you already set tnVzFilterName so here it should be 
```
vzRsSubjFiltAttAttr.TnVzFilterName = tnVzFilterName
```

And no need for the if.

File: aci/resource_aci_vzrssubjfiltatt.go (hunk 41 above)
[the remark this answers is not in the document]
Comment (lhercot) [reply aci/resource_aci_vzrssubjfiltatt.go:208]: but on line 179 you already set tnVzFilterName so here it should be 
```
vzRsSubjFiltAttAttr.TnVzFilterName = tnVzFilterName
```

And no need for the if.

File: aci/resource_aci_vzsubj.go [hunk 42]
```diff
@@ -25,6 +27,114 @@ func resourceAciContractSubject() *schema.Resource {
...
+				old, new := diff.GetChange("consumer_to_provider")
+				oldInTerm := old.(*schema.Set).List()
+				newInTerm := new.(*schema.Set).List()
+				if len(newInTerm) == len(oldInTerm) {
+					oldInTermParam := oldInTerm[0].(map[string]interface{})
+					newInTermParam := newInTerm[0].(map[string]interface{})
+
+					if oldInTermParam["prio"] != newInTermParam["prio"] || oldInTermParam["target_dscp"] != newInTermParam["target_dscp"] || oldInTermParam["relation_vz_rs_in_term_graph_att"] != newInTermParam["relation_vz_rs_in_term_graph_att"] {
+						result = true
+					} else if oldInTermParam["relation_vz_rs_filt_att"] != newInTermParam["relation_vz_rs_filt_att"] {
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:42]: isn't this condition always true as directive is a Set and is represented by an ID not by the set itself?

File: aci/resource_aci_vzsubj.go [hunk 43]
```diff
@@ -25,6 +27,114 @@ func resourceAciContractSubject() *schema.Resource {
...
+								oldInTermFilterDirectives := oldVzRsFiltAttMap["directives"].(*schema.Set).List()
+								found_same_filter_dn := false
+								for _, newInTermFilter := range newInTermFilterList {
+									newVzRsFiltAttMap := newInTermFilter.(map[string]interface{})
+									newInTermFilterDirectives := newVzRsFiltAttMap["directives"].(*schema.Set).List()
+
+									if oldVzRsFiltAttMap["filter_dn"] == newVzRsFiltAttMap["filter_dn"] {
+										found_same_filter_dn = true
+										if oldVzRsFiltAttMap["action"] != newVzRsFiltAttMap["action"] || oldVzRsFiltAttMap["priority_override"] != newVzRsFiltAttMap["priority_override"] || !reflect.DeepEqual(oldInTermFilterDirectives, newInTermFilterDirectives) {
+											if !(oldVzRsFiltAttMap["priority_override"] == "default" && newVzRsFiltAttMap["priority_override"] == "") {
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:61]: Could there be a case here where action is different but priority_override is default or empty and you ignore the change?

File: aci/resource_aci_vzsubj.go [hunk 44]
```diff
@@ -25,6 +27,114 @@ func resourceAciContractSubject() *schema.Resource {
 
 		SchemaVersion: 1,
 
+		CustomizeDiff: func(_ context.Context, diff *schema.ResourceDiff, v interface{}) error {
+			result := false
```
Comment (lhercot) [inline aci/resource_aci_vzsubj.go:31]: changed_consumer := false
changed_provider := false

**Review by lhercot — CHANGES_REQUESTED** [review]

File: go.sum [hunk 45]
```diff
@@ -77,6 +77,8 @@ github.com/chzyer/readline v0.0.0-20180603132655-2972be24d48e/go.mod h1:nSuG5e5P
 github.com/chzyer/test v0.0.0-20180213035817-a1ea475d72b1/go.mod h1:Q3SI9o4m/ZMnBNeIyt5eFwwo7qiLfzFZmjNmxjkiQlU=
 github.com/ciscoecosystem/aci-go-client v1.39.1 h1:Vc6oGDGks0Ir2M/kS33ycI9p8tu63VMVwz3YN+Y60Wc=
 github.com/ciscoecosystem/aci-go-client v1.39.1/go.mod h1:pGK3tDhR/D4+P5f6utT1V7TQQxDw2fL7j79QXeVk6po=
```
Comment (lhercot) [inline go.sum:79]: Remove old version

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**